File: src/src/main/resources/mmlspark/RecommendationIndexerModel.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class RecommendationIndexerModel(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        itemIndexModel (object): itemIndexModel
        itemInputCol (str): Item Input Col
        itemOutputCol (str): Item Output Col
        ratingCol (str): Rating Col
        userIndexModel (object): userIndexModel
        userInputCol (str): User Input Col
        userOutputCol (str): User Output Col
    """

    @keyword_only
    def __init__(self, itemIndexModel=None, itemInputCol=None, itemOutputCol=None, ratingCol=None, userIndexModel=None, userInputCol=None, userOutputCol=None):
        super(RecommendationIndexerModel, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.RecommendationIndexerModel")
        self._cache = {}
        self.itemIndexModel = Param(self, "itemIndexModel", "itemIndexModel: itemIndexModel", generateTypeConverter("itemIndexModel", self._cache, complexTypeConverter))
        self.itemInputCol = Param(self, "itemInputCol", "itemInputCol: Item Input Col")
        self.itemOutputCol = Param(self, "itemOutputCol", "itemOutputCol: Item Output Col")
        self.ratingCol = Param(self, "ratingCol", "ratingCol: Rating Col")
        self.userIndexModel = Param(self, "userIndexModel", "userIndexModel: userIndexModel", generateTypeConverter("userIndexModel", self._cache, complexTypeConverter))
        self.userInputCol = Param(self, "userInputCol", "userInputCol: User Input Col")
        self.userOutputCol = Param(self, "userOutputCol", "userOutputCol: User Output Col")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, itemIndexModel=None, itemInputCol=None, itemOutputCol=None, ratingCol=None, userIndexModel=None, userInputCol=None, userOutputCol=None):
        """
        Set the (keyword only) parameters

        Args:

            itemIndexModel (object): itemIndexModel
            itemInputCol (str): Item Input Col
            itemOutputCol (str): Item Output Col
            ratingCol (str): Rating Col
            userIndexModel (object): userIndexModel
            userInputCol (str): User Input Col
            userOutputCol (str): User Output Col
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setItemIndexModel(self, value):
        """

        Args:

            itemIndexModel (object): itemIndexModel

        """
        self._set(itemIndexModel=value)
        return self


    def getItemIndexModel(self):
        """

        Returns:

            object: itemIndexModel
        """
        return self._cache.get("itemIndexModel", None)


    def setItemInputCol(self, value):
        """

        Args:

            itemInputCol (str): Item Input Col

        """
        self._set(itemInputCol=value)
        return self


    def getItemInputCol(self):
        """

        Returns:

            str: Item Input Col
        """
        return self.getOrDefault(self.itemInputCol)


    def setItemOutputCol(self, value):
        """

        Args:

            itemOutputCol (str): Item Output Col

        """
        self._set(itemOutputCol=value)
        return self


    def getItemOutputCol(self):
        """

        Returns:

            str: Item Output Col
        """
        return self.getOrDefault(self.itemOutputCol)


    def setRatingCol(self, value):
        """

        Args:

            ratingCol (str): Rating Col

        """
        self._set(ratingCol=value)
        return self


    def getRatingCol(self):
        """

        Returns:

            str: Rating Col
        """
        return self.getOrDefault(self.ratingCol)


    def setUserIndexModel(self, value):
        """

        Args:

            userIndexModel (object): userIndexModel

        """
        self._set(userIndexModel=value)
        return self


    def getUserIndexModel(self):
        """

        Returns:

            object: userIndexModel
        """
        return self._cache.get("userIndexModel", None)


    def setUserInputCol(self, value):
        """

        Args:

            userInputCol (str): User Input Col

        """
        self._set(userInputCol=value)
        return self


    def getUserInputCol(self):
        """

        Returns:

            str: User Input Col
        """
        return self.getOrDefault(self.userInputCol)


    def setUserOutputCol(self, value):
        """

        Args:

            userOutputCol (str): User Output Col

        """
        self._set(userOutputCol=value)
        return self


    def getUserOutputCol(self):
        """

        Returns:

            str: User Output Col
        """
        return self.getOrDefault(self.userOutputCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.RecommendationIndexerModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=RecommendationIndexerModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".RecommendationIndexerModel"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/RecommendationIndexerModel.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Utils;

using SynapseML.Dotnet.Utils;


namespace Synapse.ML.Recommendation
{
    /// <summary>
    /// <see cref="RecommendationIndexerModel"/> implements RecommendationIndexerModel
    /// </summary>
    public class RecommendationIndexerModel : JavaModel<RecommendationIndexerModel>, IJavaMLWritable, IJavaMLReadable<RecommendationIndexerModel>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.recommendation.RecommendationIndexerModel";

        /// <summary>
        /// Creates a <see cref="RecommendationIndexerModel"/> without any parameters.
        /// </summary>
        public RecommendationIndexerModel() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="RecommendationIndexerModel"/> with a UID that is used to give the
        /// <see cref="RecommendationIndexerModel"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public RecommendationIndexerModel(string uid) : base(s_className, uid)
        {
        }

        internal RecommendationIndexerModel(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets itemIndexModel value for <see cref="itemIndexModel"/>
        /// </summary>
        /// <param name="itemIndexModel">
        /// itemIndexModel
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetItemIndexModel(JavaTransformer value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setItemIndexModel", (object)value));
        
        /// <summary>
        /// Sets itemInputCol value for <see cref="itemInputCol"/>
        /// </summary>
        /// <param name="itemInputCol">
        /// Item Input Col
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetItemInputCol(string value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setItemInputCol", (object)value));
        
        /// <summary>
        /// Sets itemOutputCol value for <see cref="itemOutputCol"/>
        /// </summary>
        /// <param name="itemOutputCol">
        /// Item Output Col
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetItemOutputCol(string value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setItemOutputCol", (object)value));
        
        /// <summary>
        /// Sets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <param name="ratingCol">
        /// Rating Col
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetRatingCol(string value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setRatingCol", (object)value));
        
        /// <summary>
        /// Sets userIndexModel value for <see cref="userIndexModel"/>
        /// </summary>
        /// <param name="userIndexModel">
        /// userIndexModel
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetUserIndexModel(JavaTransformer value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setUserIndexModel", (object)value));
        
        /// <summary>
        /// Sets userInputCol value for <see cref="userInputCol"/>
        /// </summary>
        /// <param name="userInputCol">
        /// User Input Col
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetUserInputCol(string value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setUserInputCol", (object)value));
        
        /// <summary>
        /// Sets userOutputCol value for <see cref="userOutputCol"/>
        /// </summary>
        /// <param name="userOutputCol">
        /// User Output Col
        /// </param>
        /// <returns> New RecommendationIndexerModel object </returns>
        public RecommendationIndexerModel SetUserOutputCol(string value) =>
            WrapAsRecommendationIndexerModel(Reference.Invoke("setUserOutputCol", (object)value));

        
        /// <summary>
        /// Gets itemIndexModel value for <see cref="itemIndexModel"/>
        /// </summary>
        /// <returns>
        /// itemIndexModel: itemIndexModel
        /// </returns>
        public JavaTransformer GetItemIndexModel()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getItemIndexModel");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets itemInputCol value for <see cref="itemInputCol"/>
        /// </summary>
        /// <returns>
        /// itemInputCol: Item Input Col
        /// </returns>
        public string GetItemInputCol() =>
            (string)Reference.Invoke("getItemInputCol");
        
        
        /// <summary>
        /// Gets itemOutputCol value for <see cref="itemOutputCol"/>
        /// </summary>
        /// <returns>
        /// itemOutputCol: Item Output Col
        /// </returns>
        public string GetItemOutputCol() =>
            (string)Reference.Invoke("getItemOutputCol");
        
        
        /// <summary>
        /// Gets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <returns>
        /// ratingCol: Rating Col
        /// </returns>
        public string GetRatingCol() =>
            (string)Reference.Invoke("getRatingCol");
        
        
        /// <summary>
        /// Gets userIndexModel value for <see cref="userIndexModel"/>
        /// </summary>
        /// <returns>
        /// userIndexModel: userIndexModel
        /// </returns>
        public JavaTransformer GetUserIndexModel()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getUserIndexModel");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets userInputCol value for <see cref="userInputCol"/>
        /// </summary>
        /// <returns>
        /// userInputCol: User Input Col
        /// </returns>
        public string GetUserInputCol() =>
            (string)Reference.Invoke("getUserInputCol");
        
        
        /// <summary>
        /// Gets userOutputCol value for <see cref="userOutputCol"/>
        /// </summary>
        /// <returns>
        /// userOutputCol: User Output Col
        /// </returns>
        public string GetUserOutputCol() =>
            (string)Reference.Invoke("getUserOutputCol");

        
        /// <summary>
        /// Loads the <see cref="RecommendationIndexerModel"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="RecommendationIndexerModel"/> was saved to</param>
        /// <returns>New <see cref="RecommendationIndexerModel"/> object, loaded from path.</returns>
        public static RecommendationIndexerModel Load(string path) => WrapAsRecommendationIndexerModel(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<RecommendationIndexerModel> Read() =>
            new JavaMLReader<RecommendationIndexerModel>((JvmObjectReference)Reference.Invoke("read"));

        private static RecommendationIndexerModel WrapAsRecommendationIndexerModel(object obj) =>
            new RecommendationIndexerModel((JvmObjectReference)obj);

        
    }
}

        